documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+- The CFG graph starting at `B1` and ending in `Bi` is a DAG (considering `fir.loop` is an
+ operation and not a CFG loop).
+
+These rules mainly mean that expression lowering does not have to (and should not) create crazy CFG in order
+to achieve short-circuiting.
+
+These rules should guarantee that we do not have to track the CFG state while lowering an expression other than the current insertion point
+(and maybe some places to load symbols. allocate temps...).
+
+### An array shape does not imply an expression is elemental
+All expressions with an array shape are not elemental, for instance `foo(array)` where `foo` is a user
+function that is not elemental cannot be replaced by call to `foo` on each `array` element.
+
+### Some non-elemental expression can be seen as an elemental expression and an affine map
+As stated in https://github.com/flang-compiler/f18/blob/master/documentation/ArrayComposition.md,
+transformational functions are not elemental but they many of them can simply be viewed as be viewed as
+changing the way their array arguments are to be accessed (i.e/ they are functions operating on the indexes and can be seen in MLIR as affine maps).
+
+### Do not early optimize but do not prevent later optimizations
+Expression lowering from the parse-tree to FIR is not meant to be an optimization pass,
+it should only produce FIR that can be optimized later. For instance, if as stated in ArrayComposition.md,
+it is an end goal to avoid unnecessary temporary arrays, lowering is not the place to perform deep temporary removal optimization.
+It is not entirely clear yet what "can be optimized later" implies.
+The strategy should be to start lowering in a simple way and if it turns out that some optimizations are later not possible, it should be
+preferred to find a way to convey the information that is missing in FIR to enable the optimization rather than to perform the optimization
+while lowering.
+
+### Not all elemental operation should be decomposed
+`merge(x, y, b)` when b is scalar does not require a loop to be implemented.
+It can be implemented as single branch to load the right value or even better, a merge
+on the memory reference.
+
+Question: If lowering were to emit a loop, would later optimization be able to transform this to a simple branch ?
+
+Note: This anyway falls in the short-circuiting cases.
+
+Also, the simple expression `array` should not be broken into a loop if this is not needed,
+it has the value `fir.load %array`.
+
+## Expression Lowering Design
+
+### Proposed Generated FIR structure for one array expression
+1. Load all required symbols (+)
+2. Compute shape (if not compiled time constant)
+3. Short-circuit if null shape
+4. Allocate tmp result (stack or heap?) (+)
+5. Compute scalar operands (+)
+6. Short-circuit if scalar operands are enough for result
+7. Compute array sub-expressions (+)
+8. Loop on element (+)
+9. short-circuit targets 
```

Comment:
That was for the CFG target block in case of short-circuits, but now that I formalized short-circuits as `fir.where`, this does not really make sens anymore.